Pod Certificate Requests: Placeholder update (#51487)
kubernetes/website: content/en/docs/reference/access-authn-authz/certificate-signing-requests.md

single PEM-encoded X.509 certificate once the CSR is approved and the certificate is issued.
Some signers store multiple certificates into the `status.certificate` field. In
that case, the documentation for the signer should specify the meaning of
additional certificates; for example, this might be the certificate plus
intermediates to be presented during TLS handshakes.

If you want to make the _trust anchor_ (root certificate) available, this should be done
separately from a CertificateSigningRequest and its `status.certificate` field. For example,
you could use a ClusterTrustBundle.

The PKCS#10 signing request format does not have a standard mechanism to specify a
certificate expiration or lifetime. The expiration or lifetime therefore has to be set
through the `spec.expirationSeconds` field of the CSR object. The built-in signers
use the `ClusterSigningDuration` configuration option, which defaults to 1 year,
(the `--cluster-signing-duration` command-line flag of the kube-controller-manager)
as the default when no `spec.expirationSeconds` is specified. When `spec.expirationSeconds`
is specified, the minimum of `spec.expirationSeconds` and `ClusterSigningDuration` is

> used.

Author: The only thing that was lost was the ClusterSigningDuration config option.  I'll fold that into the description of the built-in signers.

- mikedanese
- munnerz
- enj
title: Certificate Provisioning and Trust Distribution
api_metadata:
- apiVersion: "certificates.k8s.io/v1"
  kind: "CertificateSigningRequest"
  override_link_text: "CSR v1"
- apiVersion: "certificates.k8s.io/v1beta1"
  kind: "ClusterTrustBundle"
- apiVersion: "certificates.k8s.io/v1alpha1"
  kind: "PodCertificateRequest"  
content_type: concept
weight: 60
---

<!-- No front matter -- see https://kubernetes.io/docs/contribute/style/style-guide/#headings:~:text=Second%2Dlevel%20headings%20follow%20the%20first%2Dlevel%20(or%20the%20title)%20without%20any%20preceding%20paragraphs%20or%20texts. -->

## Signers {#signers}

Kubernetes organizes its certificate provisioning features around the concept of
a signer.

> Each signer is identified by a qualified name built from a host name and a path

Reviewer: DNS name sounds better. A hostname indicates this points to an actual host.

> Each signer is identified by a qualified name built from a host name and a path

Author: Done. 

(It seems like it might have been cleaner if signer names were *just* DNS names, but oh well).

- mikedanese
- munnerz
- enj

> title: Certificate Provisioning and Trust Distribution

Author: I agree, and I understand which content should go where.

api_metadata:
- apiVersion: "certificates.k8s.io/v1"
  kind: "CertificateSigningRequest"
  override_link_text: "CSR v1"
- apiVersion: "certificates.k8s.io/v1beta1"
  kind: "ClusterTrustBundle"
- apiVersion: "certificates.k8s.io/v1alpha1"
  kind: "PodCertificateRequest"  
content_type: concept
weight: 60
---

> <!-- No front matter -- see https://kubernetes.io/docs/contribute/style/style-guide/#headings:~:text=Second%2Dlevel%20headings%20follow%20the%20first%2Dlevel%20(or%20the%20title)%20without%20any%20preceding%20paragraphs%20or%20texts. -->

Author: OK, restored the old front matter, and slightly updated it (ClusterTrustBundles is beta now).

## Signers {#signers}

Kubernetes organizes its certificate provisioning features around the concept of
a signer.

Each signer is identified by a qualified name built from a host name and a path
segment (for example, `kubernetes.io/foo`).  Each signer uniquely identifies a
category of certificates that can be issued within the scope of a particular
cluster.  Kubernetes offers some [built-in signers](#kubernetes-signers), and
third-party signers may also be freely installed into a cluster.

Note that signers are not a concrete concept in the Kubernetes API &mdash; there
is no Signer object that you create.  Instead, several types in the

> `certificates.k8s.io` API group have a signerName field, and there is special

Author: Done.

authorization logic that lets signers handle all objects that are addressed to a
particular signer.

Each signer is backed by a set of {{< glossary_tooltip text="controllers"
term_id="controller" >}} that do one or more of the following things to let
pods, users, and external workloads provision and verify certificates issued by
the signer.
* Certificate Issuance: The signer responds to CertificateSigningRequests and
  PodCertificateRequests addressed to the signer's name.

> * Trust Anchor distribution: The controller publishes the trust anchors (root

Author: Done.

  certificates) of the signer for consumption by workloads using either
  signer-linked ClusterTrustBundles or ConfigMaps

The Kubernetes project provides a set of [built-in signers](#kubernetes-signers)
that are available in every cluster.  It is also possible to install (or build
your own) [third-party signers](#third-party-signers).

## Interacting with signers

There are three dedicated API types for workloads and users to interact with a
signer:
* CertificateSigningRequests and PodCertificateRequests allow requesting
  arbitrary certificates from a signer, and

> * ClusterTrustBundles allow users to retrieve the trust anchors (root

Author: I've added a brief description of the three types to the front matter.

certificates) of a signer.

> ### Certificate signing requests {#certificate-signing-requests}

Author: I agree, but I think I should then delete the "Interacting with signers" level 2 heading above, since it will be a stub.  Its content was already addressed in the "Signers" section.

chain and stores it in the `status.certificate` field of the CSR, encoded as a
series of PEM `CERTIFICATE` blocks. The signer implementation may, but does not
have to, consider the CSR's `spec.request` field.  It is free to issue an
entirely different certificate.

{{< note >}}
When encoded in JSON or YAML, `status.certificate` field is base-64 encoded, so
it is not directly readable.  This is handled transparently when using a client
library.
{{< /note >}}

Signer implementations are encouraged to:
* Document whether or not they consider manual approval decisions on
  CertificateSigningRequests.
* Document which parts, if any, of the PKCS#10 request they consider.
* Document the concrete format of the certificates they issue.

Once the `status.certificate` field has been populated, the request has been
completed and clients can now fetch the signed certificate PEM data from the
CSR.

If the signer controller does not wish to issue the certificate, even if it has
been approved, then it can add the `Failed` condition to the CSR, which is a
terminal state. 

Special permissions are required to sign or fail a CSR:

> * Verbs: `update`, group: `certificates.k8s.io`, resource:

Author: Done.

  `certificatesigningrequests/status`
* Verbs: `sign`, group: `certificates.k8s.io`, resource: `signers`,
  resourceName: `<signerNameDomain>/<signerNamePath>` or `<signerNameDomain>/*`

Like all conditions, the `status.conditions[].reason` field is meant to contain
a machine-readable code describing the condition in TitleCase.  The
`status.conditions[].message` field is meant for a free-form explanation for
human consumption.

In order to reduce the number of old CSR objects left in a cluster, a garbage
collection controller runs periodically. The garbage collection removes CSRs
that have not changed state for some duration:

* Approved requests: automatically deleted after 1 hour
* Denied requests: automatically deleted after 1 hour
* Failed requests: automatically deleted after 1 hour
* Pending requests: automatically deleted after 24 hours
* All requests: automatically deleted after the issued certificate has expired

> ### PodCertificateRequests {#pod-certificate-requests}

Author: Done.

## Kubernetes built-in signers {#kubernetes-signers}

Kubernetes provides some built-in signers that should be present in every
cluster.

{{< caution >}}
Signer names with domain prefix `kubernetes.io` (including any subdomains) are
reserved by the Kubernetes project.
{{< /caution >}}

### `kubernetes.io/kube-apiserver-client` {#built-in-signer-kube-apiserver-client}

This signer issues certificates that will be honored as client certificates by
the API server.

This signer responds only to CertificateSigningRequests, not
PodCertificateRequests.

There is no built-in auto-approval controller for this signer; all certificates
must be manually approved, or a third-party auto-approval controller must be
installed.

This signer does not distribute its trust bundle, since no party besides the API
server is intended to verify the certificates it issues.

When this signer's signing controller sees an approved
CertificateSigningRequest, it will check the following conditions:
* The `spec.usages` field includes at least "client auth" and at most "client auth", "digital
  signature", and "key encipherment".

If the conditions are true, the signing controller will issue the certificate.

> Otherwise it will mark the CSR as failed.

Reviewer 2: Markdown reference page per signer works for me. If we want to, we can set up a layout that renders those all into one final page (but that can come later).

> Otherwise it will mark the CSR as failed.

Author: Just noting that this discussion is still open, but the relevant content has moved to concepts/security/certificates.md

The issued certificate has the following properties:
* The X.509 Subject field Subject Alternate Names extension are copied verbatim
  from `spec.request`.
* The "Key Usage" and "Extended Key Usage" extensions are set according to the key
  usages named in `spec.usages`.
* No other extensions are included.
* NotBefore is set to the present time, with reasonable backdating to
  account for clock skew.
* NotAfter is set to the present time plus the minimum of `spec.expirationSeconds`
  and the certificate TTL configured for `kube-controller-manager` (365 days, by
  default), with reasonable backdating to account for clock skew.
* The CA bit to false.

{{< caution >}}

> Because the API server blindly treats `Subject.CommonName` and

Author: I've removed blindly, and added a specific description of what can go wrong.

standard.

This signer only responds to CertificateSigningRequests, not
PodCertificateRequests.

There is not built-in auto-approval controller for this signer.

The stable CertificateSigningRequest API (version `certificates.k8s.io/v1` and
later) does not allow to set the `signerName` as `kubernetes.io/legacy-unknown`.

This signer does not distribute its trust bundle.

When this signer's signing controller sees an approved
CertificateSigningRequest, it will immediately attempt to issue the certificate
with no further checks.

The issued certificate has the following properties:
* The X.509 Subject field Subject Alternate Names extension are copied verbatim
  from `spec.request`.
* The "Key Usage" and "Extended Key Usage" extensions are set according to the key
  usages named in `spec.usages`.
* No other extensions are included.
* NotBefore is set to the present time, with reasonable backdating to
  account for clock skew.
* NotAfter is set to the present time plus the minimum of `spec.expirationSeconds`
  and the certificate TTL configured for `kube-controller-manager` (365 days, by
  default), with reasonable backdating to account for clock skew.
* The CA bit to false.

> ## Third-Party Signer {#third-party-signers}

Author: Done.

You can also introduce your own custom signer, which should have a similar prefixed name but using your
own domain name. For example, if you represent an open source project that uses the domain `open-fictional.example`
then you might use `issuer.open-fictional.example/service-mesh` as a signer name.

To implement your custom signer, you need to provide a set of controllers that
use the Kubernetes API to interact with CertificateSigningRequests,
PodCertificateRequests, and ClusterTrustBundles that are linked to your signer's
name.

> ### Useful ClusterRoles {#authorization}

Author: I have added qualifying language to the examples.

I agree that this whole section needs to be moved to a "write your own signer" task page and be fleshed out.  Right now it is just a collection of "write your own signer" content that was scattered around this page.

> ### Useful ClusterRoles {#authorization}

Author: Just noting that this discussion is still open, but the relevant content has moved to tasks/extend-kubernetes/custom-signer.md

To distinguish them from signer-linked ClusterTrustBundles, the names of
signer-unlinked ClusterTrustBundles **must not** contain a colon (`:`).

#### Accessing ClusterTrustBundles from pods {#ctb-projection}

{{< feature-state feature_gate_name="ClusterTrustBundleProjection" >}}

The contents of ClusterTrustBundles can be injected into the container filesystem, similar to ConfigMaps and Secrets.
See the [clusterTrustBundle projected volume source](/docs/concepts/storage/projected-volumes#clustertrustbundle) for more details.

> ## Kubernetes built-in signers {#kubernetes-signers}

Author: Yeah, I was just discussing with Jordan yesterday, this page has grown so much that much of this content probably needs to be moved to a reference page, and this page should be focused on *obtaining* a client certificate for talking to `kube-apiserver`.

I would rather address that later, though, without the enhancements documentation freeze over me.

> ## Kubernetes built-in signers {#kubernetes-signers}

Reviewer 2: > _obtaining_ a client certificate for talking to kube-apiserver.

That sounds like a [task](https://kubernetes.io/docs/tasks/). https://kubernetes.io/docs/tasks/tls/ should link to the new page, ideally, even if we place it somewhere else (certificates aren't just used for TLS).

* Verbs: `sign`, group: `certificates.k8s.io`, resource: `signers`, resourceName: `<signerNameDomain>/<signerNamePath>` or `<signerNameDomain>/*`

{{% code_sample file="access/certificate-signing-request/clusterrole-sign.yaml" %}}

#### PodCertificateRequests {#useful-cluster-roles-pcr}

For PodCertificateRequests, the requester role is almost always filled by
`kubelet`, which automatically has the necessary permissions to create and read
PCRs.  There is no approver role.

Signers need to be able to watch CSRs, and issue or fail CSRs addressed to their
signer name:

* Verbs: `get`, `list`, `watch`, group: `certificates.k8s.io`, resource: `podcertificaterequests`
* Verbs: `update`, group: `certificates.k8s.io`, resource: `podcertificaterequests/status`
* Verbs: `sign`, group: `certificates.k8s.io`, resource: `signers`, resourceName: `<signerNameDomain>/<signerNamePath>` or `<signerNameDomain>/*`

{{% code_sample file="access/pod-certificate-request/clusterrole-sign.yaml" %}}

#### ClusterTrustBundles {#useful-cluster-roles-ctb}

ClusterTrustBundles have two broad roles: consumers and attesters.

Consumers need to be able to read ClusterTrustBundles.  In most scenarios, the

> consumer will be a node identity or service account, both of which automatically

Author: Hmm, nodes get their authorization through the node authorizer, not RBAC.  But I will clarify.

> consumer will be a node identity or service account, both of which automatically

Author: Done.

have permission to read all ClusterTrustBundles.
* Verbs: `get`, `list`, `watch`, group `certificates.k8s.io`, resource: `clustertrustbundles`

Attesters are typically a signer controller, and will need permission to create
and maintain specific signer-linked ClusterTrustBundles

> * Verbs: `create`, `get`, `list`, `watch`, group: `certificates.k8s.io`, resource: `clustertrustbundles`

Author: Done.

The contents of ClusterTrustBundles can be injected into the container filesystem, similar to ConfigMaps and Secrets.
See the [clusterTrustBundle projected volume source](/docs/concepts/storage/projected-volumes#clustertrustbundle) for more details.

## Kubernetes built-in signers {#kubernetes-signers}

Kubernetes provides some built-in signers that should be present in every
cluster.

> {{< caution >}}

Reviewer 2: This only needs to be a note.

Signer names with domain prefix `kubernetes.io` (including any subdomains) are
reserved by the Kubernetes project.
{{< /caution >}}

### `kubernetes.io/kube-apiserver-client` {#built-in-signer-kube-apiserver-client}

This signer issues certificates that will be honored as client certificates by
the API server.

This signer responds only to CertificateSigningRequests, not
PodCertificateRequests.

> There is no built-in auto-approval controller for this signer; all certificates
> must be manually approved, or a third-party auto-approval controller must be
> installed.

Reviewer 2: ```suggestion
There is no built-in auto-approval controller for this signer; all certificates
that specify the `kubernetes.io/kube-apiserver-client` signer must either be
manually approved, or be approved automatically by a (third party)
{{< glossary_tooltip text="controller" term_id="controller" >}}.
```

group, but this is typically not the only highly-privileged identity present in
a cluster.

{{</ caution >}}

### `kubernetes.io/kube-apiserver-client-kubelet` {#built-in-signer-kube-apiserver-client-kubelet}

This signer issues certificates that will be honored as `kubelet` client
certificates by the API server.  It offers no further guarantees on what the
certificates may be used for.

This signer responds only to CertificateSigningRequests, not
PodCertificateRequests.

There is a built-in auto-approval controller for this signer.

This signer does not distribute its trust bundle, since no party besides the API
server is intended to verify the certificates it issues.

When this signer's signing controller sees an approved
CertificateSigningRequest, it will check the following conditions:
* The `Subject.CommonName` field has the prefix `system:node:`
* There is one `Subject.Organization` entry, exactly equal to `system:nodes`
* There are no entries in the Subject Alternate Name extension
* The CSR `spec.usages` field is either:
  * `digital signature` and `client auth`, or
  * `digital signature`, `client auth`, and `key encipherment`. 

> 

Reviewer 2: Do we check the CA flag in the CSR?

If the conditions are true, the signing controller will issue the certificate.
Otherwise it will mark the CSR as failed.

The issued certificate has the following properties:
* The X.509 Subject field Subject Alternate Names extension are copied verbatim
  from `spec.request`.
* The "Key Usage" and "Extended Key Usage" extensions are set according to the key
  usages named in `spec.usages`.
* No other extensions are included.
* NotBefore is set to the present time, with reasonable backdating to
  account for clock skew.
* NotAfter is set to the present time plus the minimum of `spec.expirationSeconds`
  and the certificate TTL configured for `kube-controller-manager` (365 days, by
  default), with reasonable backdating to account for clock skew.

> * The CA bit to false.

Reviewer 2: ```suggestion
* The CA flag is false.
```

### `kubernetes.io/kubelet-serving` {#built-in-signer-kube-apiserver-client-kubelet-serving}

This signer issues certificates that the API server will honor as valid
`kubelet` serving certificates.  It offers no further guarantees on what the
certificates may be used for.

This signer responds only to CertificateSigningRequests, not
PodCertificateRequests.

There is no built-in auto-approval controller for this signer.

This signer does not distribute its trust bundle, since no party besides the API
server is intended to verify the certificates it issues.

When this signer's signing controller sees an approved
CertificateSigningRequest, it will check the following conditions:
* The `Subject.CommonName` field has the prefix `system:node:`
* There is one `Subject.Organization` entry, exactly equal to `system:nodes`
* There is at least one DNS or IP entry in Subject Alternate Name extension.
* There are no Email or URI entries in the Subject Alternate Name extension.

> * The CSR `spec.usages` field is either:

Reviewer 2: For clarity, consider:
```suggestion
* The CertificateSigningRequest object's `spec.usages` field is either:
```

  * `digital signature`, `server auth`, and `key encipherment`. 

If the conditions are true, the signing controller will issue the certificate.
Otherwise it will mark the CSR as failed.

The issued certificate has the following properties:
* The X.509 Subject field Subject Alternate Names extension are copied verbatim
  from `spec.request`.
* The "Key Usage" and "Extended Key Usage" extensions are set according to the key
  usages named in `spec.usages`.
* No other extensions are included.
* NotBefore is set to the present time, with reasonable backdating to
  account for clock skew.
* NotAfter is set to the present time plus the minimum of `spec.expirationSeconds`
  and the certificate TTL configured for `kube-controller-manager` (365 days, by
  default), with reasonable backdating to account for clock skew.
* The CA bit to false.

### `kubernetes.io/kube-apiserver-serving` {#built-in-signer-kube-apiserver-serving}

{{< feature-state feature_gate_name="ClusterTrustBundle" >}}

This signer does not issue any certificates.  It purely serves to distribute the
trust anchors that allow workloads within the cluster to verify the API server's

> server certificate when they connect to `kubernetes.default.svc`.

Reviewer 2: ```suggestion
server certificate when they connect to the `kubernetes` Service in the default
namespace.
```

AFAIK we don't guarantee that `kubernetes.default.svc` resolves, and even if it's part of conformance, there are actual clusters where it might not.

There are three built-in distribution mechanisms for these trust anchors:
* A built-in controller writes them into the `.data[ca.crt]` field of
  ServiceAccount token secrets.

> * A built-in controller creates a ConfigMap named `kube-root-ca.crt` in each

Reviewer 2: Is this true? Even system namespaces like `kube-public`?

>   namespace.  The `.data["ca.crt"]` field of this configmap contains the trust

Reviewer 2: ```suggestion
  namespace.  The `.data["ca.crt"]` field of this ConfigMap contains the trust
```

  anchors of this signer.
* (Only when ClusterTrustBundles are enabled) A built-in controller creates a
  set of signer-linked ClusterTrustBundles for this signer.  Workloads can
  access them directly, or using `clusterTrustBundle` projected volumes sources
  (`signerName="kubernetes.io/kube-apiserver-serving"` and
  `labelSelector.matchLabels={}`).

### `kubernetes.io/legacy-unknown` {#built-in-signer-legacy-unknown}

This signer issues certificates that are not guaranteed to be suitable for any
purpose.  In some third-party distributions of Kubernetes, the API server may
accept certificates it issues as client certificates, but this is not a
standard.

This signer only responds to CertificateSigningRequests, not
PodCertificateRequests.

There is not built-in auto-approval controller for this signer.

> The stable CertificateSigningRequest API (version `certificates.k8s.io/v1` and
> later) does not allow to set the `signerName` as `kubernetes.io/legacy-unknown`.

Reviewer 2: ```suggestion
{{< note >}}
The stable CertificateSigningRequest API (version `certificates.k8s.io/v1` and
later) does not allow you to set the `signerName` as `kubernetes.io/legacy-unknown`.
{{< /note >}}
```

> 

Reviewer 2: Do we still support the older (pre-v1) APIs? If not, any associated CSRs might be considered legacy.

This signer does not distribute its trust bundle.

When this signer's signing controller sees an approved
CertificateSigningRequest, it will immediately attempt to issue the certificate
with no further checks.

The issued certificate has the following properties:
* The X.509 Subject field Subject Alternate Names extension are copied verbatim
  from `spec.request`.
* The "Key Usage" and "Extended Key Usage" extensions are set according to the key
  usages named in `spec.usages`.
* No other extensions are included.
* NotBefore is set to the present time, with reasonable backdating to
  account for clock skew.
* NotAfter is set to the present time plus the minimum of `spec.expirationSeconds`
  and the certificate TTL configured for `kube-controller-manager` (365 days, by
  default), with reasonable backdating to account for clock skew.

> * The CA bit to false.

Reviewer 2: ```suggestion
* The CA flag is false.
```

consumer will be `kubelet` (automatic permission via the node authorizer, if
enabled) or a service account (automatic permission via an RBAC bootstrap
ClusterRole, if enabled).
* Verbs: **get**, **list**, **watch**, group `certificates.k8s.io`, resource: `clustertrustbundles`

Attesters are typically a signer controller, and will need permission to create
and maintain specific signer-linked ClusterTrustBundles
* Verbs: **create**, **get**, **list**, **watch**, group: `certificates.k8s.io`, resource: `clustertrustbundles`
* Verbs: **attest**, group: `certificates.k8s.io`, resource: `signers`, resourceName: `<signerNameDomain>/<signerNamePath>` or `<signerNameDomain>/*`

### Writing a PodCertificateRequest controller {#write-pcr-controller}

Under the hood, `kubelet` runs the state machine depicted in Figure 1 for each
`podCertificate` projection.  The actions your signer takes on the
PodCertificateRequests that `kubelet` generates control the state transitions.

{{< mermaid >}} 
stateDiagram-v2

direction LR

Initial --> Wait
Wait --> Fresh
Wait --> Failed
Wait --> Denied
Fresh --> WaitRefresh
WaitRefresh --> Failed
WaitRefresh --> Denied

> 

Reviewer 2: nit: we prefer using Kubernetes colors (see https://kubernetes.io/docs/contribute/style/diagram-guide/)
